### DISCUSSION TITLE
How to reset form from outside the form?

### DISCUSSION THREAD

**Discussion by msueping** [react-hook-form/react-hook-form#6728] [Q&A] [answered]
I am displaying a react-hook-form (having an id="myForm") _inside of a Material-ui V5 dialog_. Initially I had two buttons within the form itself that submitted or reset the form data. That worked fine for me.

Now I'm trying to move the two buttons and their functionality to the dialog's footer (i.e., the initial form buttons are now higher up in the hierarchy; at the dialog level. Hope that makes sense). I am able to get the form to submit successfully by specifying the new Submit button's props for type="submit" and form="myForm". What is not working for me is trying to do something similar for the Reset button. Setting its type="reset" along with form="myForm" does not do anything.

The original Reset button (that was part of the actual form) had onClick={() => methods.reset()} which worked. 

How can I make the Reset button that is now in the dialog behave the same as it did when it was part of the actual form? For what it's worth, below is a screen shot of the two buttons that are part of the dialog's footer and the actual form is just above the dividing line

![image](https://user-images.githubusercontent.com/67658122/136222507-ae0d8c28-bc74-4d59-b3a0-39c4ba784c56.png)

**Comment by thanh-nguyen-95** [comment, marked as answer]
If I understand it correctly, your problem something like this:

```jsx
/* before */
<form onSubmit={/* ... */}>
  <input />
  <button type="submit">Submit button</button> /* this works */
  <button type="reset" onClick={() => method.reset()}>Reset button</button> /* this works */
</form>

/* after */
<form id="myForm" onSubmit={/* ... */}>
  <input />
</form>

<button type="submit" form="myForm">Submit button</button> /* this works */
<button type="reset" form="myForm">Reset button</button> /* this does not work */
```

I think the reset button worked because it triggered the _native_ reset feature of the browser (clearing all input values inside a form), not react-hook-form's.

My suggestion is to make the reset button like this:

```jsx
/* remove the type="reset" and add a click handler */
<button onClick={() => method.reset()}>Reset button</button>
```

If that doesn't work, try passing a new form value to `reset()`:

```jsx
/* this will 'replace' your entire form value (not a 'merge') so be careful */
<button
  onClick={() => method.reset({
    name: '',
    age: 18,
    job: null,
  })}
>
  Reset button
</button>
```

Please provide a simple CodeSandBox so we can help you better. I hope this helps.

**Reply by msueping** [reply]
I managed to get it to work after taking your suggestions into consideration. Thank you.

The following codesandbox is what I've come up with. I'm not sure if I've coded it the best way but it seems to work. There is a small issue however. If you open the dialog and edit the name from "Bob" to "Bobby" or anything else then choose "Save", the console log shows that the textbox's data got updated, however, the form data that was initially used is still saying "Bob". If you choose "Save" again (without changing the name and leaving it as is), you'll see the changed name now showing in both sets of data. Any idea what's causing that?

https://codesandbox.io/s/formdialog-material-demo-forked-e5dkg?file=/public/MyForm.js

**Reply by bluebill1049** [reply]
thanks @thanh-nguyen-95 for the detailed answer 👍

**Reply by msueping** [reply]
Hi bluebill1049, did you happen to look at my codesandbox? Is that the correct approach that you'd recommend? Do you have any idea what would be causing the small issue that I've observed regarding the values of the data when submitting?

**Reply by thanh-nguyen-95** [reply]
> 
> 
> Hi bluebill1049, did you happen to look at my codesandbox? Is that the correct approach that you'd recommend? Do you have any idea what would be causing the small issue that I've observed regarding the values of the data when submitting?

`setFormData` is asynchronous so when you call `console.log(formData)` it's not updated yet. If you want to check if a React state is changed, use `useEffect` to check it. For example:

```js
useEffect(() => {
  console.log(formData)
}, [formData])
```

**Reply by msueping** [reply]
OK. That makes sense.

What about my overall approach to handling the submit and reset from the dialog's buttons? Is there a better way to do that?

**Reply by thanh-nguyen-95** [reply]
It really depends on your business logic and how you setup your components. My advice is to make everything works first, then come back and optimize later. That way you can have an accurate view of the whole application.

The code looks good to me so I don't have anything to say about that.
